osd/scrub: extracting scrub "backend" functionality

ronen-fr opened the pull request:
Moved scrub backend functionality, i.e. (more or less) all functions that check and manipulate
h-objects actual data, into a separate class.

The relevant code is extracted from (mostly) the PG, PrimaryLogPG, PgScrubber, PrimaryLogScrub
and PgBackend.

rzarzynski commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+
+void ScrubQueue::dec_scrubs_remote()
+{
+  dout(20) << __func__ << " " << m_scrubs_remote << " -> " << (m_scrubs_remote - 1)
```
`m_scrubs_remote` is accessed outside the critical section. It's a logging code but this would introduce another degree of freedom for a developer during an investigation: _does the text in the file fit the internal state?_

ronen-fr replied:
Not a clear-cut decision, as there is a benefit in shortening the time under lock, but I think you are right.  Will move the log back under the lock.

rzarzynski reviewed (commented):
Reviewed `osd_scrub_sched.h` and `osd_scrub_sched.cc` from af587b5de381013e8f568960f3fc5e43d4e8b51c. Generally this looks really good; the code is high quality. Mostly questions / suggestions / minors / nits.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+{
+  int hb_interval = m_cct->_conf->osd_heartbeat_interval;
+  int n_samples = 60 * 24 * 24;
+  if (hb_interval > 1) {
```
Agree. That's actually the original code (and I plan to post an existing rewrite of the whole way it's done),
but I'll change that anyway.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+// ////////////////////////////////////////////////////////////////////////////////// //
+// ScrubJob
+
+#define dout_context (cct)
```
Agree. Result of multiple fixes. Will fix.

ronen-fr replied:
In up-to-date code there is a log in that range.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+  }
+
+  // get CPU load avg
+  double loadavgs[1];
```
will change.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+
+class PG;
+
+namespace Scrub {
```
Already using it in existing files. Will handle that in a separate PR.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+enum class attempt_t {
+  scrubbing,	    // successfully started a scrub
+  none_ready,	    // no pg to scrub
+  local_resources,  // failure to secure local OSD scrub resource
```
Hmm.. To me it's clearer that way: "failed in reserving the *local resources*". Don't have any
real attachment to this, anyway.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+  scrubbing,	    // successfully started a scrub
+  none_ready,	    // no pg to scrub
+  local_resources,  // failure to secure local OSD scrub resource
+  already_started,  // already started scrubbing this pg
```
Will rephrase the comment here. Or did you think of locations where the 'already_started' is used?

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+
+  ScrubQueue(CephContext* cct, OSDService& osds);
+
+  struct TimeAndDeadline {
```
changed (per Kefu's suggestion) in the original PR to scrub_schedule_t

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+    /// the OSD id (for the log)
+    int whoami;
+
+    std::atomic<qu_state_t> m_state{qu_state_t::created};
```
Fixed here.
Didn't yet fix in the other three atomic_bool instances. Will add this to ceph_atomic.h separately.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+    utime_t m_sched_time;
+
+    /// the hard upper bound of scrub time
+    utime_t m_deadline;
```
Good point. Will fix (unless a good reason is found...)

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+
+  double daily_loadavg{0.0};
+
+  struct time_indirect_t {
```
Already removed. Will update with the up-to-date state of https://github.com/ceph/ceph/pull/40984/

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+    utime_t get_sched_time() const { return m_sched_time; }
+
+    /// relatively low-cost(*) access to the scrub-job's state, to be used in logging.
+    /// (*) not on x64 architecture
```
My worry was regarding the atomic access to the index, not regarding the translation into strings.
On X64 the "relaxed" access used here for debugging is no cheaper than the full S.C. one.

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+   *
+   * locking: locks jobs_lock
+   */
+  Scrub::attempt_t select_pg_and_scrub(Scrub::ScrubPreconds& preconds);
```
Modified the comment. 'preconds' is a "convenience" collection of preconditions, only some of which are known
when select_pg_and_scrub() is called. Ostensibly the parameter should have been by-copy (which is costly),
or I could have just passed the known parameters separately (which was less convenient).

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+    case qu_state_t::pg_deleted: return "PG removed"sv;
+  }
+  // g++ (unlike CLANG), requires an extra 'return' here
+  return "(unknown)"sv;
```
On second thought: not sure I understand what you meant here?

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+    utime_t suggested_stamp{};
+    double min_interval{0.0};
+    double max_interval{0.0};
+    must_scrub_t is_must{ScrubQueue::must_scrub_t::not_mandatory};
```
'must' is the term used in the original (pre my time) scrub code whenever we refer to these conditions

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.h:
```diff
@@ -0,0 +1,331 @@
+  double daily_loadavg{0.0};
+
+  struct time_indirect_t {
+    bool operator()(const ceph::ref_t<ScrubJob>& lhs,
```
all fixed

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+    case qu_state_t::pg_deleted:
+      // RRR understand why the PG wasn't fully removed
+      dout(5) << " Restoring a scrub job" << dendl;
+      // fall through
```
not there anymore

ronen-fr commented on src/osd/scrubber/osd_scrub_sched.cc:
```diff
@@ -0,0 +1,721 @@
+  for (auto jb = to_scrub.begin(); jb != to_scrub.end();) {
+
+    if ((*jb)->m_resources_failure) {
+
```
dropped one of the two... The other helps me see the structure.

ronen-fr commented:
Added embedded documentation (ASCII art)

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+
+using shard_info_map_t = std::map<pg_shard_t, shard_info_wrapper>;
+
+using IterToSMap = std::map<pg_shard_t, ScrubMap>::iterator;
```
using shard_to_scrubmap_t = td::map<pg_shard_t, ScrubMap>;

Just use shard_to_scrubmap_t::iterator directly.

ronen-fr replied:
done

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+ *
+ * Created by the Scrubber after all replicas' maps have arrived.
+ */
+struct ScrubBeChunk {
```
scrub_chunk_t to match other names in this file.

athanatos replied:
ScrubBackend::new_chunk() exposes this entire structure to PGScrubber.  I'm not sure what we're trying to encapsulate here.

ronen-fr replied:
> scrub_chunk_t to match other names in this file.

Followed the guidance that very simple constructs should be named with a '_t', while the
rest using Camel. Can change if that's the pref.

athanatos commented on src/osd/scrubber/pg_scrubber.h:
```diff
@@ -3,6 +3,70 @@
+                                 │
+                                 │
+┌────────────────────────────────▼──────────────────┐
+│             <<ScrubPgIf>>                         │
```
ScrubPgIF

ronen-fr replied:
<< >> seems to be the recommended UML notation?

athanatos replied:
Sorry, I mean that the 'f' in ScrubPgIF is lower case rather than upper case as it is in the code.

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+   * sets Backend's m_repair flag (setting m_mode_desc to a corresponding
+   * string)
+   */
+  void update_repair_status(bool should_repair) final;
```
This only seems to get called during PGScrubber::scrub_finish.  I'm not sure I understand the intent.

ronen-fr replied:
Basically - a copy of PgScrubber's mode text and 'repairing' status. Created so that the backend won't have to query 'backwards' for these details.

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+  void update_repair_status(bool should_repair) final;
+
+  // note: used by both Primary & replicas
+  static ScrubMap clean_meta_map(ScrubMap& cleaned, bool max_reached);
```
Not used externally, should probably not be public.

athanatos commented on src/osd/scrubber/pg_scrubber.cc:
```diff
@@ -1113,10 +1038,8 @@ int PgScrubber::build_replica_map_chunk()
-      auto for_meta_scrub = clean_meta_map();
-      _scan_snaps(for_meta_scrub);
+
+      m_be->replica_clean_meta(replica_scrubmap, m_end.is_max(), m_start);
```
This should probably be part of build_scrub_map_chunk()?

ronen-fr replied:
That will be a change in order of operations for the Primary. Maybe can be done - but surely not for this PR.

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+    return num_digest_updates_pending;
+  }
+
+  void scan_snaps(ScrubMap& smap) final;
```
Not used externally, shouldn't be public.

ronen-fr replied:
fixed

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk() final;
```
It seems like PGScrubber::build_scrub_map() should probably return a ScrubMap that then gets passed here to something like ScrubBackend::add_primary_map.

ronen-fr replied:
Was this way. Changed when I noticed that by having the primary map as just one more entry in the collected set of maps, I can avoid creating a "set of pointers to maps, which are all but one in the same vector" and then having to access the maps via this indirection (there is a comment somewhere to that effect).

athanatos replied:
Seems like add_primary_map could still add to this set?

ronen-fr replied:
Will create an extra copy. Prefer not to do that if no reason.

ronen-fr replied:
Good point... will revise

ronen-fr replied:
@athanatos - done

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,419 @@
+                                      std::set<pg_shard_t>    ///< object_errors
+                                      >;
+
+  std::optional<ScrubBeChunk> this_chunk;
```
new_chunk() has a side effect of clearing this_chunk.  It's not clear to me where m_missing or m_inconsistent get cleared.  I don't really understand why PGScrubber doesn't just create a new instance of this entire class for each chunk -- there doesn't seem to be any state intended to survive between chunks.

ronen-fr replied:
I forgot to rm the old m_missing (BTW - clearing was not its problem... it was just cleared, with no other usage). Thanks for spotting that.

The whole scrub backend is created anew each scrub.

athanatos commented:
I just left an initial round of comments.  As far as I can tell, the purpose of ScrubBackend is to encapsulate the process of comparing the maps from all replicas (including the primary) and reporting errors.  Perhaps it should be called ScrubChunkAnalyzer or ScrubChunkChecker?

ScrubBackendIF seems entirely unnecessary.  The behavior of the implementation is entirely constrained by the overall constraints on scrub, so I don't anticipate another implementation in the future.  The interface also doesn't enable adding a testing shim as far as I can tell.  Perhaps I'm missing something.

I'd really like to see unit tests for this logic.  The main obstacles at the moment seem to be the m_scrubber, m_pgbe, and m_pg members.  Let's have a look at the users:
* ScrubBackend::process_clones_to: accesses m_shallow_errors via m_scrubber
* ScrubBackend::record_object_digests: uses m_scrubber, m_pg to get an object context and submit a write
* ScrubBackend::scan_object_snaps uses m_pg.snap_mapper to get local indexed snaps
* ScrubBackend::repair_oinfo_oid: Add method to return std::vector<std::pair<hobject_t, object_info_t>> of objects that need object infos repaired and do the actual repair in PGScrubber
* ScrubBackend::repair_object: we use m_pg to get to force_object_missing.  This could also be part of PGScrubber since we apparently
* ...


These users also mean that this refactor doesn't really make this logic any easier to use from crimson since we're combining the general logic for comparing maps with the classic specific logic for performing IO.  I'd rather see the public methods of ScrubBackend (ScrubChunkAnalyzer) return structs describing the IO that need to be done so that PGScrubber can actually perform them.  You'd be able to easily add unit tests as there wouldn't be any external dependencies and it would render a large chunk of business logic usable from crimson.  As it is, I'm not sure what this refactor actually buys us.

ronen-fr commented:
@athanatos 
>>  It seems like we're combining the general logic for comparing maps with the classic specific logic for performing IO.

Yes - that's a fare description. in my thinking, both functions (the comparing and the resulting I/O operations) are "below" the PgScrubber in the hierarchy. I see what you are saying, and will try to come up with a modified solution. At first glance, I'd rather break the ScrubberBE into the two components - the maps vs the I/O. Not sure I'd like to refer the I/O back "upwards" to PgScrubber.

athanatos commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -0,0 +1,1957 @@
+  return &this_chunk->received_maps[m_pg_whoami];
+}
+
+void ScrubBackend::merge_to_master_set()
```
Sorry for not noticing this earlier.  We're trying to avoid adding any further usages of the word 'master' to the code base. Let's use authoritative rather than master here -- it's more descriptive anyway.

ronen-fr replied:
done

athanatos commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -0,0 +1,1957 @@
+    }
+  }
+
+  // Construct the master set of objects
```
authoritative

athanatos commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -0,0 +1,1957 @@
+  // construct authoritative scrub map for type-specific scrubbing
+
+  m_cleaned_meta_map.insert(my_map());
+  merge_to_master_set();
```
merge_to_authoritative_set

athanatos commented on src/osd/scrubber/scrub_backend.cc:
```diff
@@ -0,0 +1,1957 @@
+// re-implementation of PGBackend::be_compare_scrubmaps()
+std::optional<std::string> ScrubBackend::compare_smaps()
+{
+  dout(10) << __func__ << ": master-set #: " << this_chunk->master_set.size()
```
authoritative_set

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,402 @@
+  std::map<pg_shard_t, ScrubMap> received_maps;
+
+  /// a collection of all objs mentioned in the maps
+  std::set<hobject_t> master_set;
```
authoritative_set

ronen-fr replied:
done

athanatos commented on src/osd/scrubber/scrub_backend.h:
```diff
@@ -0,0 +1,402 @@
+   *   - this_chunk->maps[from] with the replicas' scrub-maps;
+   *   - this_chunk->master_set as a union of all the maps' objects;
+   */
+  void merge_to_master_set();
```
merge_to_authoritative_set (and comments above)

ronen-fr replied:
done

ronen-fr commented:
@athanatos
> It seems like we're combining the general logic for comparing maps with the classic specific logic for performing IO. I'd rather
> see the public methods of ScrubBackend (ScrubChunkAnalyzer) return structs describing the IO that needs to be done so 
> that PGScrubber can actually perform them. You'd be able to easily add unit tests as there wouldn't be any external 
> dependencies and it would render a large chunk of business logic usable from crimson. As it is, I'm not sure what this refactor 
> actually buys us since ScrubBackend has essentially the same dependencies.

Based on our conversation, I am treating this PR as a 1`st phase (with the 2'nd to be discussed before being finalized).
I did have to incorporate one of the changes you have suggested now: as the ungainly 'context op' backwards interface from the Backend to the PrimaryLogPG I have added resulted in an '#include' issue, it is now removed:
ScrubBackend::scrub_snapshot_metadata() now submits a vector of digest fixes upwards.

This removes some of the extra interfaces you have mentioned. Most of the rest will be handled when teh ScrubStore is
factored-out.

ronen-fr commented:
Run 100 relevant teuthology tests  (https://pulpito.ceph.com/?branch=wip-ronenf-scrub-be2) on 8f591ab.
One failure (https://pulpito.ceph.com/rfriedma-2022-01-13_14:57:26-rados:standalone-wip-ronenf-scrub-be2-distro-basic-smithi/6612712/). Seems unrelated. Checking

ronen-fr commented:
jenkins test make check

athanatos commented:
retest this please

athanatos commented:
retest this please

ronen-fr commented:
failed 'make check' test seems related to https://github.com/ceph/ceph/pull/42112 (not related to this PR).
@rzarzynski  - can you take a look? you handled that change

ronen-fr commented:
retest this please

ronen-fr commented:
retest this please

ronen-fr commented:
Retested on Teuthology again on 22.1.22

wjwithagen commented:
Clang is not really happy on FreeBSD.
I guess that the refactoring is the cause.
```
/home/jenkins/workspace/ceph-master-compile/src/osd/scrubber/pg_scrubber.h:854:35: error: no matching constructor for initialization of 'std::lock_guard<std::mutex>'
      std::lock_guard<std::mutex> lk{m_preemption_lock};
                                  ^ ~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__mutex_base:89:14: note: candidate constructor not viable: no known conversion from 'ceph::mutex' (aka 'mutex_debug_impl<false>') to 'std::lock_guard<std::mutex>::mutex_type &' (aka 'std::mutex &') for 1st argument
    explicit lock_guard(mutex_type& __m) _LIBCPP_THREAD_SAFETY_ANNOTATION(acquire_capability(__m))
             ^
/usr/include/c++/v1/__mutex_base:99:5: note: candidate constructor not viable: no known conversion from 'ceph::mutex' (aka 'mutex_debug_impl<false>') to 'const std::lock_guard<std::mutex>' for 1st argument
    lock_guard(lock_guard const&) _LIBCPP_EQUAL_DELETE;
    ^
/usr/include/c++/v1/__mutex_base:93:5: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
    lock_guard(mutex_type& __m, adopt_lock_t) _LIBCPP_THREAD_SAFETY_ANNOTATION(requires_capability(__m))
    ^
In file included from /home/jenkins/workspace/ceph-master-compile/src/osd/scrubber/osd_scrub_sched.cc:9:
/home/jenkins/workspace/ceph-master-compile/src/osd/scrubber/pg_scrubber.h:868:35: error: no matching constructor for initialization of 'std::lock_guard<std::mutex>'
      std::lock_guard<std::mutex> lk{m_preemption_lock};
                                  ^ ~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__mutex_base:89:14: note: candidate constructor not viable: no known conversion from 'ceph::mutex' (aka 'mutex_debug_impl<false>') to 'std::lock_guard<std::mutex>::mutex_type &' (aka 'std::mutex &') for 1st argument
    explicit lock_guard(mutex_type& __m) _LIBCPP_THREAD_SAFETY_ANNOTATION(acquire_capability(__m))
             ^
```

For previous experience it is probably that:
    `std::lock_guard<std::mutex> lk{m_preemption_lock};`
does not agree with the definition:
    `mutable ceph::mutex m_preemption_lock;`
    
Now I I only knew where I had that before.

ronen-fr commented:
Noted. Will start working on a fix.

wjwithagen commented:
Thanx

> Noted. Will start working on a fix. And @wjwithagen - thank you.

Thanx. Now If I only knew what the previous case was...
I did not fix that either, so can't find it in my PRs. :-(

ronen-fr commented:
@wjwithagen - does https://github.com/ceph/ceph/pull/44755 (there are 4-5 lines changed) solve the compilation problem?
That PR is approved, but there are some unrelated 'make check' issues that delay its merge.
